pkg/audit/manager.go

```diff
@@ -363,34 +387,37 @@ func (am *Manager) getUpdateListsFromAuditResponses(res []*constraintTypes.Resul
 		message := r.Msg
 		resource, ok := r.Resource.(*unstructured.Unstructured)
 		if !ok {
-			return nil, nil, nil, errors.Errorf("could not cast resource as reviewResource: %v", r.Resource)
+			return errors.Errorf("could not cast resource as reviewResource: %v", r.Resource)
 		}
 		rname := resource.GetName()
 		rkind := resource.GetKind()
 		rnamespace := resource.GetNamespace()
-		result := auditResult{
-			cgvk:              gvk,
-			capiversion:       apiVersion,
-			cname:             name,
-			cnamespace:        namespace,
-			rkind:             rkind,
-			rname:             rname,
-			rnamespace:        rnamespace,
-			message:           message,
-			enforcementAction: enforcementAction,
-			constraint:        r.Constraint,
+		// append audit results only if it is below violations limit
+		if uint(len(updateLists[selfLink])) < *constraintViolationsLimit {
+			result := auditResult{
+				cgvk:              gvk,
+				capiversion:       apiVersion,
+				cname:             name,
+				cnamespace:        namespace,
+				rkind:             rkind,
+				rname:             rname,
+				rnamespace:        rnamespace,
+				message:           message,
+				enforcementAction: enforcementAction,
+				constraint:        r.Constraint,
+			}
+			updateLists[selfLink] = append(updateLists[selfLink], result)
```

Comment:
Instead of calling `writeAuditResults` at the very end of the whole audit, we could call `writeAuditResults` here as soon as `totalViolationsPerConstraint[selfLink] == *constraintViolationsLimit` is reached to update that constraint. then we can remove updateLists[selfLink] from memory as we no longer need to persist it.

Reply:
Can we follow up with this as a further optimization? added https://github.com/open-policy-agent/gatekeeper/issues/711#issuecomment-662708692

Reply:
+1 on waiting to see the impact of this change before weighing whether the added complexity is worth the benefit.

Reply:
sgtm